File: app/src/main/java/com/app/turingrobot/ui/fragment/chat/ChatFragment.java
package com.app.turingrobot.ui.fragment.chat;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputLayout;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatButton;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.app.turingrobot.R;
import com.app.turingrobot.common.Constants;
import com.app.turingrobot.core.App;
import com.app.turingrobot.entity.CoreEntity;
import com.app.turingrobot.ui.adapter.ChatAdapter;
import com.app.turingrobot.ui.core.BaseFragment;
import com.app.turingrobot.ui.dialog.AuthDialogFragment;
import com.app.turingrobot.utils.TUtil;
import com.socks.library.KLog;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * 聊天
 * Created by Alpha on 2016/3/26 21:56.
 */
public class ChatFragment extends BaseFragment implements SwipeRefreshLayout.OnRefreshListener, View.OnClickListener {


    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;

    @BindView(R.id.swipeRefresh)
    SwipeRefreshLayout swipeRefresh;

    @BindView(R.id.inputLayoutExtra)
    TextInputLayout inputLayoutExtra;

    @BindView(R.id.btn_send)
    AppCompatButton btnSend;

    private ChatAdapter chatAdapter;

    private Disposable mDisp;

    private Unbinder unbind;


    public static ChatFragment newInstance() {
        ChatFragment chatFragment = new ChatFragment();

        return chatFragment;
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_chat, container, false);
        unbind = ButterKnife.bind(this, view);

        init();

        return view;
    }

    private void init() {
        swipeRefresh.setOnRefreshListener(this);
        btnSend.setOnClickListener(this);
        inputLayoutExtra.setHint("");
        inputLayoutExtra.setHint("Message");

        chatAdapter = new ChatAdapter(activity);
        LinearLayoutManager layoutManager = new LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(chatAdapter);

        sendInfo("Hello!");

    }

    /**
     * 刷新
     */
    @Override
    public void onRefresh() {

    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_send:

                if (App.getUser() == null) {
                    AuthDialogFragment.newInstance()
                            .show(activity.getFragmentManager(), AuthDialogFragment.class.getSimpleName());
                    return;
                }

                String msg = inputLayoutExtra.getEditText().getText().toString().trim();
                if (TextUtils.isEmpty(msg)) {
                    TUtil.show("请输入信息...");
                    return;
                }
                inputLayoutExtra.getEditText().setText("");
                CoreEntity coreEntity = new CoreEntity();
                coreEntity.setText(msg);
                coreEntity.setTime(System.currentTimeMillis());
                coreEntity.setTarget(false);
                chatAdapter.addData(coreEntity, recyclerView);
                sendInfo(msg);
                break;
        }
    }

    /**
     * 请求数据
     *
     * @param msg
     */
    private void sendInfo(String msg) {
        mDisp = apiService.getText(Constants.APP_KEY, msg)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(val -> {
                    val.setTime(System.currentTimeMillis());
                    chatAdapter.addData(val, recyclerView);
                }, KLog::e);
    }


    @Override
    public void onDestroyView() {
        unbind.unbind();
        if (mDisp != null) {
            mDisp.dispose();
        }
        super.onDestroyView();
    }

}

File: app/src/main/java/com/app/turingrobot/ui/fragment/chat/ChatFragment.kt
package com.app.turingrobot.ui.fragment.chat

import android.os.Bundle
import android.os.Handler
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.app.turingrobot.R
import com.app.turingrobot.app.App
import com.app.turingrobot.entity.CoreEntity
import com.app.turingrobot.helper.rx.RxResultHelper
import com.app.turingrobot.multitype.MultiTypeAdapter
import com.app.turingrobot.multitype.base.MultiTypePresenterImpl
import com.app.turingrobot.multitype.base.ViewModel
import com.app.turingrobot.ui.core.BaseFragment
import com.app.turingrobot.ui.dialog.AuthDialogFragment
import com.app.turingrobot.ui.fragment.chat.model.LinkModel
import com.app.turingrobot.ui.fragment.chat.model.TextMsgModel
import com.app.turingrobot.ui.fragment.chat.model.TextMsgTargetModel
import com.app.turingrobot.utils.TUtil
import com.socks.library.KLog
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.fragment_chat.*
import java.util.ArrayList

/**
 * 聊天
 * Created by Alpha on 2016/3/26 21:56.
 */
class ChatFragment : BaseFragment(), SwipeRefreshLayout.OnRefreshListener, View.OnClickListener {

    private var mAdapter: MultiTypeAdapter? = null

    private val mDisp = CompositeDisposable()

    private val mData: MutableList<ViewModel> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater?.inflate(R.layout.fragment_chat, container, false)
        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        init()
    }

    private fun init() {

        swipeRefresh.setOnRefreshListener(this)
        btn_send.setOnClickListener(this)

        inputLayoutExtra.hint = "Message"

        val multiTypePresenter = MultiTypePresenterImpl
                .newBuilder()
                .addHolders(TextMsgModel.newPresenter())
                .addHolders(TextMsgTargetModel.newPresenter())
                .addHolders(LinkModel.newPresenter())
                .withViewModels(mData)
                .build()

        mAdapter = MultiTypeAdapter(multiTypePresenter)

        val layoutManager = LinearLayoutManager(activity,
                LinearLayoutManager.VERTICAL, false)
        recyclerView.layoutManager = layoutManager
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.setHasFixedSize(true)
        recyclerView.adapter = mAdapter

        sendInfo("Hello")

    }

    /**
     * 刷新
     */
    override fun onRefresh() {

        Handler().postDelayed({
            swipeRefresh.isRefreshing = false
        }, 2000)

    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_send -> {

                if (App.getUser() == null) {
                    AuthDialogFragment.newInstance()
                            .show(activity?.fragmentManager,
                                    AuthDialogFragment::class.java.simpleName)
                    return
                }

                val msg = inputLayoutExtra!!.editText!!.text.toString().trim { it <= ' ' }
                if (TextUtils.isEmpty(msg)) {
                    TUtil.show("请输入信息...")
                    return
                }

                inputLayoutExtra!!.editText!!.setText("")
                val coreEntity = CoreEntity()
                coreEntity.text = msg
                mData.add(TextMsgModel(coreEntity))
                mAdapter?.notifyItemInserted(mData.size - 1)
                recyclerView.scrollToPosition(mData.size - 1)

                sendInfo(msg)
            }
        }
    }

    /**
     * 请求数据
     * @param msg
     */
    private fun sendInfo(msg: String) {
        mDisp.add(apiService.sendMsg("2a397396709f674d8996787b5b0b0b12", msg)
                .compose(RxResultHelper.transform())
                .subscribe({
                    mData.add(it)
                    mAdapter?.notifyItemInserted(mData.size - 1)
                    recyclerView.scrollToPosition(mData.size - 1)
                }, KLog::e))
    }


    override fun onDestroyView() {
        mDisp.dispose()
        super.onDestroyView()
    }

    companion object {

        fun newInstance(): ChatFragment {
            return ChatFragment()
        }
    }

}
